### PR TITLE
[ML] Wait for .ml-config primary before assigning persistent tasks

### PR DESCRIPTION
**Pull request by droberts195**
Now that ML job configs are stored in an index rather than
cluster state, availability of the .ml-config index is very
important to the operation of ML.  When a cluster starts up
the ML persistent tasks will be considered for node
assignment very early on.  It is best in this case if
assignment is deferred until after the .ml-config index is
available.

The introduction of data frame analytics jobs has made this
problem worse, because anomaly detection jobs already waited
for the primary shards of the .ml-state, .ml-anomalies-shared
and .ml-meta indices to be available before doing node
assignment, and by coincidence this would probably lead to
the primary shards of .ml-config also being searchable.  But
data frame analytics jobs had no other index checks prior to
this change.  This is why I propose to only apply this fix to 7.3
and above.

This fixes problem 2 of #44156

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDataFrameAnalyticsAction.java [hunk 1]
```diff
@@ -422,6 +440,14 @@ protected AllocatedPersistentTask createTask(
 
             String id = params.getId();
 
+            List<String> unavailableIndices = verifyIndicesPrimaryShardsAreActive(clusterState, AnomalyDetectorsIndex.configIndexName());
+            if (unavailableIndices.size() != 0) {
+                String reason = "Not opening job [" + id + "], because not all primary shards are active for the following indices [" +
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDataFrameAnalyticsAction.java]: Do we want to use the word `job` here? Do you think we should say `data frame analytics job` and then the ID? Just to keep it clear it is not an anomaly job?

**Review by benwtrent — APPROVED** [review]
One minor comment around logging message.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDataFrameAnalyticsAction.java (hunk 1 above)
Comment (droberts195) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDataFrameAnalyticsAction.java]: Yes, good point.  The two types of jobs are in different namespaces so the message should say which it is.  I'll change the equivalent message in the other file too.

Comment (droberts195) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDataFrameAnalyticsAction.java]: > I'll change the equivalent message in the other file too.

On reflection I think changing "job" to "anomaly detection job" in existing error messages should be done in a different PR.  There are many places that would need changing to make it consistent.

**Pull request merged**